packages/opentelemetry-core/src/utils/deep-merge.ts

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+export function deepMerge(
+  target: Record<string, any>,
+  source: Record<string, any>,
+  maxDepth = 10
+) {
+  const merged = target;
+  if (maxDepth === 0) {
+    throw new Error('Max depth exceeded.');
+  }
+  for (const [prop, value] of Object.entries(source)) {
+    if (bothPropsAreArrays(target, source, prop)) {
+      merged[prop] = [];
+      merged[prop] = value;
```

Comment:
why do you set it first as an array and then override it with value ? If the last line 
`merged[prop] = value;`
should win then this condition is not needed as it is exactly the same action as `else` so you could simplify it to 
```
if (bothPropsAreObjects(target, source, prop)) {
} else {
}
```
unless this is a bug and should be fixed differently.

Reply:
Not sure about this one, as it stands the `merged[prop]` at that time holds the target array as a value, but we want it to hold the source object's value however (as this util will replace arrays entirely). i.e. we want to replace its array value entirely, that's why i reset it first, unless there is a function to replace arrays I am unaware of?

Reply:
but you are assigning to `merged[prop]`  2 different values - one after another so it doesn't make sense to overrride it first with array (`[]`) and then with `value`

Reply:
Ah yes thanks. the initialization does seem unnecessary. I thought `[4]` merged into `[1,2,3]` would result in `[4,2,3]` without the initialization but does not seem to be the case.

Reply:
fixed, PTAL

Reply:
ok so now just last concern, you have here 3 conditions

1. `if (bothPropsAreArrays(target, source, prop)) {`
2. `} else if (bothPropsAreObjects(target, source, prop)) {`
3. `else`

Condition 1 and 3 returns the same, so as I mentioned at the beginning you can simplify it to
```
if (bothPropsAreObjects(target, source, prop)) {
} else {
}
```

Reply:
ah right, so I tried that, in in fact both props being arrays does satisfy bothPropsAreObjects, in that case we do get the [4] merged into [1,2,3] resulting in [4,2,3] .

The alternative would be to have the function be  `bothPropsAreObjectsButNotArrays` or something less verbose that depicts that

Reply:
Do you prefer a `bothPripsAreObjectsButNotArrays` approach or the order-being-important 3 case approach as is now?

Reply:
a ok then that makes sense now if arrays are being detected also as an objects, I would leave it as it is then, just add comment that the first condition is needed t prevent array being parsed as objects, thx

Reply:
I just refactored the if statement to make it apparent without comment, PTAL